### PR TITLE
Migrate codeBlockButton plugins to TypeScript

### PR DESCRIPTION
**Pull request by jovyntls**
**What is the purpose of this pull request?**

- [ ] Documentation update
- [ ] Bug fix
- [ ] Feature addition or enhancement
- [x] Code maintenance
- [ ] DevOps
- [ ] Improve developer experience
- [ ] Others, please explain:

Part of #1913 

**Overview of changes:**
Migrate codeBlockButtonsContainer to TS

**Anything you'd like to highlight/discuss:**
Should we standardise a type for `node` that combines its possible types (e.g. `DomElement & cheerio.Element`)? I'm not sure if it's necessary (would having these different types for nodes cause issues as we continue the migration?).

**Testing instructions:**
N/A

**Proposed commit message: (wrap lines at 72 characters)**
(merge commit)

---

**Checklist:** :ballot_box_with_check:

- [ ] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes

### CODE REVIEW

**Comment by raysonkoh** [comment]
>  Should we standardise a type for node that combines its possible types (e.g. DomElement & cheerio.Element)? I'm not sure if it's necessary (would having these different types for nodes cause issues as we continue the migration?).

I think it's a good idea, we could maybe have our own type `MarkbindNode` (ideally immutable), and have some functions that wrap the common methods that we call on `DomElement` or `cheerio.Element`.

My thoughts are to do the migration fully first, before implementing this as this is potentially a breaking change. Also there are not enough unit tests for the core package that I'm confident that it will detect any subtle bugs introduced. Wdyt?

**Comment by jovyntls** [comment]
> I think it's a good idea, we could maybe have our own type `MarkbindNode` (ideally immutable), and have some functions that wrap the common methods that we call on `DomElement` or `cheerio.Element`.

Yeah agree with having an interface for Node! Standardising all the common methods would make the codebase more readable and less confusing, especially since we have code that looks similar with significant differences (e.g. `cheerio.html(node)` vs `cheerio(node).html`

> My thoughts are to do the migration fully first, before implementing this as this is potentially a breaking change. Also there are not enough unit tests for the core package that I'm confident that it will detect any subtle bugs introduced. Wdyt?

My thoughts:
- Creating the types first could help with the TS migration by making it easier to type `node` arguments/parameters. We may also be able to reduce the number of checks for undefined/null attributes if we have a type for nodes.
- I agree that having more unit tests would good, since we may end up introducing some unintended effects with such a large scale migration
- Also, if we further refactor the code to use this new Node interface/type, the changes might be so significant that our git history detects the `.ts` file as a new file rather than a renaming of a previous `.js` file

**Comment by tlylt** [comment]
> My thoughts:
> 
> * Creating the types first could help with the TS migration by making it easier to type `node` arguments/parameters. We may also be able to reduce the number of checks for undefined/null attributes if we have a type for nodes.
> * I agree that having more unit tests would good, since we may end up introducing some unintended effects with such a large scale migration
> * Also, if we further refactor the code to use this new Node interface/type, the changes might be so significant that our git history detects the `.ts` file as a new file rather than a renaming of a previous `.js` file

Jumping in with some brief comments since I have not done any TS migration myself:
- will be good to create an issue to keep track of the TODO
- in the issue, also would be good to mention an expedient solution when encountering the type of DomElement vs cheerio.Element for the moment so that we can deal with this uniformly (as much as possible).
  - `DomElement & cheerio.Element` seems ok, perhaps @raysonkoh can give a second opinion on this
- upon checking, `cheerio` seems to have migrated to TypeScript with the [latest `v1` release](https://github.com/cheeriojs/cheerio/releases/tag/v1.0.0-rc.12), might be useful to take a look there
  - could be related to #2075

File: packages/core/src/plugins/codeBlockCopyButtons.ts
Comment (raysonkoh) [inline packages/core/src/plugins/codeBlockCopyButtons.ts:51]: Interesting use of intersection types!

**Review by raysonkoh — APPROVED** [review]
LGTM

File: packages/core/src/plugins/codeBlockCopyButtons.ts
Comment (tlylt) [inline packages/core/src/plugins/codeBlockCopyButtons.ts:51]: one small question: is `pluginContext` marked as `any` here because it's a user defined json and hard to type?

Comment (jovyntls) [reply packages/core/src/plugins/codeBlockCopyButtons.ts:51]: I marked it as `any` because it's not used anywhere in the function. Not sure if `_:any` or `_: PluginContext` would be better?

Comment (ong6) [reply packages/core/src/plugins/codeBlockCopyButtons.ts:51]: why not just use `_:PluginContext` since we're only taking that in anyways

Comment (jovyntls) [reply packages/core/src/plugins/codeBlockCopyButtons.ts]: Thanks for the input guys, created a `PluginsContext` class in types.ts - let me know what you think

File: packages/core/src/plugins/types.ts
Comment (raysonkoh) [inline packages/core/src/plugins/types.ts:1]: Nit: can we place this in an existing file instead of creating a new file? Maybe `Plugin.js`?

Comment (jovyntls) [reply packages/core/src/plugins/types.ts:1]: Just left a [comment](https://github.com/MarkBind/markbind/issues/2114#issuecomment-1413805848) on #2114 related to this - my concern was that putting types in an existing file could create cyclic dependency issues. We shouldn't have cyclic dependencies in this case for plugins, but if we choose to standardise how the types are handled, a `types.ts` may be a solution. What do you think? For now I could leave it in Plugin.js and see how it goes?

Comment (tlylt) [reply packages/core/src/plugins/types.ts:1]: Perhaps can also consider using the `d.ts` extension? Somehow `types.ts` feels like an anti-pattern to me.

Comment (raysonkoh) [reply packages/core/src/plugins/types.ts:1]: I think as long as we standardize how we organize the types, it _should_ be fine. I would prefer something more descriptive than `types.ts` though. 

> For now I could leave it in Plugin.js and see how it goes?

Yup! I think you can do this for now. Let's finish up with migration first since it is higher priority and we can come back later to reorganize the types.

Comment (jovyntls) [reply packages/core/src/plugins/types.ts:1]: Speaking of Plugin.js - should we put this PR on hold until plugin.js is migrated, since the types will be defined there?

Comment (tlylt) [reply packages/core/src/plugins/types.ts:1]: I think merging this in first is fine, can refactor it later if needed.

**Review by tlylt — APPROVED** [review]

**Pull request merged**
